### PR TITLE
[feature] add paging to AP following / followers endpoints

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
# Description
- adds paging to the AP following endpoint: `/users/{username}/following`
- adds paging to the AP followers endpoint: `/users/{username}/followers`
- adds new ActivityStreams Collection (like) utility builder functions

Thankfully, a much shorter paging PR. https://github.com/superseriousbusiness/gotosocial/pull/2186 walked so #2198 could run.

## Checklist
- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [ ] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/ap/activitystreams.go [hunk 1]
```diff
@@ -77,3 +86,197 @@ const (
...
+	limit := "limit=" + strconv.Itoa(pageLimit)
+	firstIRI.RawQuery = appendQuery(firstIRI.RawQuery, limit)
+
+	// Add the collection first IRI property.
+	first := streams.NewActivityStreamsFirstProperty()
+	first.SetIRI(firstIRI)
+	collection.SetActivityStreamsFirst(first)
+}
+
+func buildCollectionPage[C CollectionPageBuilder, I ItemsPropertyBuilder](collectionPage C, itemsProp I, setItems func(I), params CollectionPageParams) {
```
Comment (tsmethurst) [inline internal/ap/activitystreams.go:204]: Oh damn okay now that's a function signature!

File: internal/ap/activitystreams.go [hunk 2]
```diff
@@ -77,3 +86,197 @@ const (
...
+	)
+
+	if nextIRI != nil {
+		// Add the collection next property for the next page.
+		nextProp := streams.NewActivityStreamsNextProperty()
+		nextProp.SetIRI(nextIRI)
+		collectionPage.SetActivityStreamsNext(nextProp)
+	}
+
+	// Build the prevocab.page link IRI.
```
Comment (tsmethurst) [inline internal/ap/activitystreams.go:239]: prevocab :D Is this a find + replace artifact?

**Review by tsmethurst — COMMENTED** [review]
Looking slick so far :eyes:

File: internal/ap/activitystreams.go (hunk 1 above)
Comment (NyaaaWhatsUpDoc) [reply internal/ap/activitystreams.go:204]: yeah it's not the prettiest, but massively improves code reuse and it's only the package private part of it :D

File: internal/processing/fedi/collections.go [hunk 3]
```diff
@@ -128,24 +196,90 @@ func (p *Processor) FollowersGet(ctx context.Context, requestedUsername string)
...
 	if errWithCode != nil {
 		return nil, errWithCode
 	}
 
-	requestedAccountURI, err := url.Parse(requestedAccount.URI)
+	// Parse the collection ID object from account's following URI.
+	collectionID, err := url.Parse(requestedAccount.FollowingURI)
 	if err != nil {
-		return nil, gtserror.NewErrorInternalError(fmt.Errorf("error parsing url %s: %s", requestedAccount.URI, err))
+		err := gtserror.Newf("error parsing account following uri %s: %v", requestedAccount.FollowingURI, err)
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/processing/fedi/collections.go:208]: brain fart, is the reason ;p

File: internal/ap/activitystreams.go (hunk 2 above)
Comment (NyaaaWhatsUpDoc) [reply internal/ap/activitystreams.go:239]: maybe... what's it to ya

File: internal/processing/fedi/collections.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/processing/fedi/collections.go]: (fixed :blush: )

File: internal/ap/activitystreams.go (hunk 2 above)
Comment (NyaaaWhatsUpDoc) [reply internal/ap/activitystreams.go]: (fixed :blush: )

**Comment by tsmethurst** [comment]
Looks good!!

**Review by tsmethurst — APPROVED** [review]

**Pull request merged**